pkg/gardenlet/controller/controllerinstallation/controllerinstallation/reconciler_test.go

```diff
@@ -0,0 +1,134 @@
+// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package controllerinstallation_test
+
+import (
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	appsv1 "k8s.io/api/apps/v1"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/utils/ptr"
+
+	. "github.com/gardener/gardener/pkg/gardenlet/controller/controllerinstallation/controllerinstallation"
+)
+
+var _ = Describe("Reconciler", func() {
+	var (
+		reconciler *Reconciler
+	)
+
+	BeforeEach(func() {
+		reconciler = &Reconciler{}
+	})
+
+	Describe("#BootstrapControlPlaneNodeFunc", func() {
+		var (
+			pod         *corev1.Pod
+			deployment  *appsv1.Deployment
+			statefulSet *appsv1.StatefulSet
+			daemonSet   *appsv1.DaemonSet
+			job         *batchv1.Job
+			cronJob     *batchv1.CronJob
+		)
+
+		BeforeEach(func() {
+			pod = &corev1.Pod{}
+			deployment = &appsv1.Deployment{}
+			statefulSet = &appsv1.StatefulSet{}
+			daemonSet = &appsv1.DaemonSet{}
+			job = &batchv1.Job{}
+			cronJob = &batchv1.CronJob{}
+		})
+
+		It("should not change objects if not bootstrapping control plane node", func() {
+			reconciler.BootstrapControlPlaneNode = false
+
+			p := pod.DeepCopy()
+			Expect(reconciler.BootstrapControlPlaneNodeFunc(p)).To(Succeed())
+			Expect(p).To(Equal(pod))
+
+			d := deployment.DeepCopy()
+			Expect(reconciler.BootstrapControlPlaneNodeFunc(d)).To(Succeed())
+			Expect(d).To(Equal(deployment))
+
+			s := statefulSet.DeepCopy()
+			Expect(reconciler.BootstrapControlPlaneNodeFunc(s)).To(Succeed())
+			Expect(s).To(Equal(statefulSet))
+
+			ds := daemonSet.DeepCopy()
+			Expect(reconciler.BootstrapControlPlaneNodeFunc(ds)).To(Succeed())
+			Expect(ds).To(Equal(daemonSet))
+
+			j := job.DeepCopy()
+			Expect(reconciler.BootstrapControlPlaneNodeFunc(j)).To(Succeed())
+			Expect(j).To(Equal(job))
+
+			c := cronJob.DeepCopy()
+			Expect(reconciler.BootstrapControlPlaneNodeFunc(c)).To(Succeed())
+			Expect(c).To(Equal(cronJob))
+		})
+
+		It("should adapt objects if bootstrapping control plane node", func() {
+			reconciler.BootstrapControlPlaneNode = true
+
+			Expect(reconciler.BootstrapControlPlaneNodeFunc(pod)).To(Succeed())
+			Expect(pod.Spec.HostNetwork).To(BeTrue())
+			Expect(pod.Spec.Tolerations).To(Equal([]corev1.Toleration{
+				{Operator: corev1.TolerationOpExists, Effect: corev1.TaintEffectNoSchedule},
+				{Operator: corev1.TolerationOpExists, Effect: corev1.TaintEffectNoExecute},
+			}))
+
+			Expect(reconciler.BootstrapControlPlaneNodeFunc(deployment)).To(Succeed())
+			Expect(deployment.Spec.Replicas).To(Equal(ptr.To(int32(1))))
+			Expect(deployment.Spec.Template.Spec.HostNetwork).To(BeTrue())
```

Comment:
Can you extract a common matcher instead of repeating the host network and tolerations check for every type?
The `HaveField` matcher could help with this, as the `Spec.Template.Spec.HostNetwork` field path is the same for multiple object kinds, even though it traverses different types.